File: scripts/_vllm_plugins.py
"""vLLM parser-plugin registry — Python side.

Reads ``deploy/vllm-plugins.json`` (the single source of truth, also
consumed by ``gpu-arbiter/main.go``) and exposes simple lookup helpers
for the probe driver. The registry maps a parser name (the value passed
to ``--tool-call-parser`` / ``--reasoning-parser``) to the plugin file
that registers it. Names absent from the registry are treated as
built-in vLLM parsers and pass through unchanged.

Adding a new plugin: drop the ``.py`` file in ``scripts/vllm_plugins/``
and add one entry in ``deploy/vllm-plugins.json``. Nothing else needs
to change in this module.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
DEFAULT_REGISTRY_PATH = REPO_ROOT / "deploy" / "vllm-plugins.json"
DEFAULT_HOST_PLUGIN_DIR = REPO_ROOT / "scripts" / "vllm_plugins"
DEFAULT_CONTAINER_DIR = "/etc/devai/vllm-plugins"


@dataclass(frozen=True)
class PluginEntry:
    name: str
    kind: str  # "tool" | "reasoning"
    file: str  # basename inside the plugins dir

    def container_path(self, container_dir: str) -> str:
        return f"{container_dir.rstrip('/')}/{self.file}"


@dataclass(frozen=True)
class PluginRegistry:
    container_dir: str
    host_dir: Path
    entries: dict[str, PluginEntry]

    def lookup(self, parser_name: str | None) -> PluginEntry | None:
        if not parser_name:
            return None
        return self.entries.get(parser_name)

    def is_plugin(self, parser_name: str | None) -> bool:
        return self.lookup(parser_name) is not None


_CACHED_REGISTRY: PluginRegistry | None = None


def get_registry(
    path: Path | None = None,
    *,
    host_dir: Path | None = None,
    refresh: bool = False,
) -> PluginRegistry:
    """Module-level cache. Within a probe pass the registry file does
    not change; loading it once avoids re-parsing on every cell.
    """
    global _CACHED_REGISTRY
    if refresh or _CACHED_REGISTRY is None:
        _CACHED_REGISTRY = load_registry(path, host_dir=host_dir)
    return _CACHED_REGISTRY


def load_registry(
    path: Path | None = None,
    *,
    host_dir: Path | None = None,
) -> PluginRegistry:
    """Load the JSON registry. Missing file or malformed JSON yields an
    empty registry — every parser name then routes through the built-in
    path (no plugin flag, no bind-mount). This keeps the probe and
    router behaviour identical to pre-plugin builds when the registry
    is absent.
    """
    p = path or DEFAULT_REGISTRY_PATH
    container_dir = DEFAULT_CONTAINER_DIR
    entries: dict[str, PluginEntry] = {}
    try:
        raw = json.loads(p.read_text())
    except (FileNotFoundError, json.JSONDecodeError):
        raw = {}
    if isinstance(raw, dict):
        cd = raw.get("container_dir")
        if isinstance(cd, str) and cd.strip():
            container_dir = cd.strip()
        plugins = raw.get("plugins")
        if isinstance(plugins, dict):
            for name, meta in plugins.items():
                if not isinstance(name, str) or not isinstance(meta, dict):
                    continue
                kind = meta.get("kind")
                fname = meta.get("file")
                if kind not in ("tool", "reasoning"):
                    continue
                if not isinstance(fname, str) or not fname.strip():
                    continue
                entries[name] = PluginEntry(
                    name=name, kind=kind, file=fname.strip()
                )
    resolved_host_dir = host_dir or Path(
        os.environ.get("VLLM_PLUGINS_HOST_DIR")
        or DEFAULT_HOST_PLUGIN_DIR
    )
    return PluginRegistry(
        container_dir=container_dir,
        host_dir=resolved_host_dir,
        entries=entries,
    )

File: gpu-arbiter/vllm_plugins.go
package main

// vLLM parser-plugin registry — Go side.
//
// Reads deploy/vllm-plugins.json (the same file consumed by
// scripts/_vllm_plugins.py) and resolves parser names to in-container
// plugin file paths. When a model's tool_parser or reasoning_parser
// name matches a registry entry, the router (a) bind-mounts the
// host plugin directory into the recreated backend container at the
// registry's container_dir, and (b) prepends the corresponding
// `--*-parser-plugin <abs-path>` flag in front of the parser-name
// flag (vLLM resolves parser names at the point the parser-name flag
// is evaluated, so the plugin module must be loaded by then).
//
// Names absent from the registry are treated as built-in vLLM parsers
// and pass through unchanged — the router's behaviour for built-ins
// is identical to pre-plugin builds.
//
// Adding a new plugin: drop the .py file in scripts/vllm_plugins/ and
// add one entry in deploy/vllm-plugins.json. Nothing in this file
// needs to change.

import (
	"encoding/json"
	"log"
	"os"
	"strings"
)

const defaultPluginContainerDir = "/etc/devai/vllm-plugins"

type vllmPluginEntry struct {
	Kind string `json:"kind"` // "tool" | "reasoning"
	File string `json:"file"` // basename inside the plugins dir
}

type vllmPluginRegistry struct {
	ContainerDir string                     `json:"container_dir"`
	Plugins      map[string]vllmPluginEntry `json:"plugins"`
	// HostDir is populated from VLLM_PLUGINS_HOST_DIR (not from the JSON
	// file) so the same registry can be shared across hosts with
	// different repo locations. Empty when the env var is unset; in
	// that mode any model that requires a plugin will fail at
	// containerRecreate time with an actionable error rather than
	// silently launching without the plugin.
	HostDir string `json:"-"`
}

// loadVLLMPluginRegistry reads the JSON registry. A missing file or
// parse error logs a warning and returns an empty (non-nil) registry —
// every parser name then routes through the built-in path, matching
// pre-plugin behaviour.
func loadVLLMPluginRegistry(path, hostDir string) *vllmPluginRegistry {
	r := &vllmPluginRegistry{
		ContainerDir: defaultPluginContainerDir,
		Plugins:      map[string]vllmPluginEntry{},
		HostDir:      hostDir,
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			log.Printf("vllm plugin registry: %s not present — built-in parsers only", path)
		} else {
			log.Printf("warning: vllm plugin registry %s read failed: %v", path, err)
		}
		return r
	}
	// The registry tolerates an `_comment` field at the top level (used
	// for human-readable notes); only `container_dir` and `plugins` are
	// consumed.
	var raw struct {
		ContainerDir string                     `json:"container_dir"`
		Plugins      map[string]vllmPluginEntry `json:"plugins"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		log.Printf("warning: vllm plugin registry %s parse failed: %v", path, err)
		return r
	}
	if strings.TrimSpace(raw.ContainerDir) != "" {
		r.ContainerDir = strings.TrimSpace(raw.ContainerDir)
	}
	for name, entry := range raw.Plugins {
		if entry.Kind != "tool" && entry.Kind != "reasoning" {
			log.Printf("warning: vllm plugin %q has invalid kind=%q; skipping", name, entry.Kind)
			continue
		}
		if strings.TrimSpace(entry.File) == "" {
			log.Printf("warning: vllm plugin %q missing file; skipping", name)
			continue
		}
		r.Plugins[name] = entry
	}
	log.Printf("vllm plugin registry: %s loaded (%d entries; host_dir=%q)",
		path, len(r.Plugins), r.HostDir)
	return r
}

// Lookup returns the registered entry for parserName plus whether it
// matched. A nil receiver and an empty parserName both yield (_, false).
func (r *vllmPluginRegistry) Lookup(parserName string) (vllmPluginEntry, bool) {
	if r == nil || parserName == "" {
		return vllmPluginEntry{}, false
	}
	e, ok := r.Plugins[parserName]
	return e, ok
}

// ContainerPath joins the registry's container_dir with the plugin
// file's basename, producing the absolute path inside the backend
// container that gets passed to `--tool-parser-plugin` /
// `--reasoning-parser-plugin`.
func (r *vllmPluginRegistry) ContainerPath(file string) string {
	if r == nil {
		return ""
	}
	dir := strings.TrimRight(r.ContainerDir, "/")
	if dir == "" {
		dir = strings.TrimRight(defaultPluginContainerDir, "/")
	}
	return dir + "/" + file
}
